Created lil edi for usaco 861 (#6346)
cpinitiative/usaco-guide: solutions/gold/usaco-861.mdx

about which haybale each cow should choose, let us think about what makes a haybale valid.
Suppose a cow starts at pasture $i$ and decides to eat at haybale $h$. Since she will
eventually travel from $h$ to the barn optimally, the total length of her route is going to be:

$$
dist(i,h) + d_h
$$

where $d_v$ represents the shortest distance from the pasture $v$ to the barn.

Normally, the cow would travel to the barn in $d_i$ time. Thus, the extra distance being
caused by the haybale $h$ would be $dist(i,h) + d_h - d_i$. This haybale would be beneficial
only if this extra distance is at most its yumminess, i.e  $dist(i,h) + d_h - d_i \leq y_h$.

If we were to rearrange this inequality, we get $dist(i,h) + d_h - y_h \leq d_i$.  Now, notice that the left-hand side of our inequality,

$$
dist(i,h) + d_h - y_h,
$$

already resembles a shortest-path distance. The only difference is that instead of starting with distance $0$,
each haybale $h$ starts with an initial distance of $d_h - y_h$. Therefore, we can
initialize every haybale with distance $d_h - y_h$, insert all haybales into the priority queue, and run a

> single **multi-source Dijkstra**. This computes $\min_h \left(dist(i,h)+d(h)-y(h)\right)$ for every pasture

Reviewer: Can you justify why ms dijkstra is used?

Author: because there are more than 1 haybales?

id: usaco-861
source: USACO Gold 2018 December
title: Fine Dining
author: Abhinav Singh, Nathan Gong
---

[Official Analysis (C++)](http://www.usaco.org/current/data/sol_dining_gold_dec18.html)

## Explanation

As each cow walks home, and along the way she's allowed exactly one detour to a haybale, but
only if the detour doesn't eat up more extra time than that haybale's yumminess is worth.
We need to decide, for every cow, whether *some* haybale is worth the detour.

The brute-force version checks every (cow, haybale) pair: is the detour cheap enough?
However, that's $O(NK)$ shortest-path checks, which is infeasible. Instead of thinking
about which haybale each cow should choose, let us think about what makes a haybale valid.

> Suppose a cow starts at pasture $i$ and decides to eat at haybale $h$. Since she will

Reviewer: ```suggestion

Suppose a cow starts at pasture $i$ and decides to eat at haybale $h$. Since she will
```

eventually travel from $h$ to the barn optimally, the total length of her route is going to be:

$$
dist(i,h) + d_h
$$

where $d_v$ represents the shortest distance from the pasture $v$ to the barn.

Normally, the cow would travel to the barn in $d_i$ time. Thus, the extra distance being
caused by the haybale $h$ would be $dist(i,h) + d_h - d_i$. This haybale would be beneficial
only if this extra distance is at most its yumminess, i.e  $dist(i,h) + d_h - d_i \leq y_h$.

If we were to rearrange this inequality, we get $dist(i,h) + d_h - y_h \leq d_i$.  Now, notice that the left-hand side of our inequality,

$$
dist(i,h) + d_h - y_h,
$$

already resembles a shortest-path distance. The only difference is that instead of starting with distance $0$,
each haybale $h$ starts with an initial distance of $d_h - y_h$. Therefore, we can
initialize every haybale with distance $d_h - y_h$, insert all haybales into the priority queue, and run a
single **multi-source Dijkstra**. We use multi-source Dijkstra because we need the minimum value over all haybales:
This computes $\min_h \left(dist(i,h)+d(h)-y(h)\right)$ for every pasture

> $i$ simultaneously. Finally, for each pasture $i$, we compare the value computed by the multi-source Dijkstra

Reviewer: ```suggestion
single **multi-source Dijkstra**. We use multi-source Dijkstra because running seperate Dijkstra from each haybale would be too slow $(\mathcal{O}(K (N + M) \log N))$ to process;  
instead, if we start from all haybales at once, we would be able to compute $\min_h \left(dist(i,h)+d(h)-y(h)\right)$ for every pasture
$i$ in $\mathcal{O}((N + M) \log N)$. Finally, for each pasture $i$, we compare the value computed by the multi-source Dijkstra
```

with $d_i$. If it is at most $d_i$, then there exists a haybale whose detour is no greater than its yumminess,
so we output 1; otherwise, we output 0.

## Implementation

> **Time Complexity:** $\mathcal{O}(M\log N)$

Reviewer: ```suggestion
**Time Complexity:** $\mathcal{O}((N + M) \log N)$
```

id: usaco-861
source: USACO Gold 2018 December
title: Fine Dining
author: Abhinav Singh, Nathan Gong
---

[Official Analysis (C++)](http://www.usaco.org/current/data/sol_dining_gold_dec18.html)

## Explanation

As each cow walks home, and along the way she's allowed exactly one detour to a haybale, but

> only if the detour doesn't eat up more extra time than that haybale's yumminess is worth.

Reviewer 2: this is kind of nitpicky, but most guide articles are written in a style similar to math/technical writing

e.g. 
> As each cow walks home, and along the way she's allowed exactly one detour to a haybale, bu only if the detour doesn't eat up more extra time than that haybale's yumminess is worth.

might be better written as 
> As each cow walks home, it can consume a haybale if the yumminess is greater than or equal to the extra distance walked to reach the haybale.

<CPPSection>

```cpp

#include <bits/stdc++.h>
using namespace std;

const long long INF = 1e18;

> int main() {

Reviewer: if you are changing c++ solution then change it for java also. but i dont understand what is the point of changing the solution

Author: the old soln wasn't matching my edi

Reviewer: that's not a reasonable excuse. can you change your editorial based on previous soln. If you have another way of solving it then either include in user solution or add new header - Solution 2

Author: There was no previous sol and imo my code is simpler than his
